devito/dle/blocking_utils.py

```diff
@@ -134,43 +139,46 @@ def optimize_unfolded_tree(unfolded, root):
     """
     processed = []
     for i, tree in enumerate(unfolded):
+        assert len(tree) == len(root)
         otree = []
-        stmts = []
+        oroot = []
         mapper = {}
 
         # "Shrink" the iteration space
-        for j in tree:
-            start, end, incr = j.args['limits']
-            otree.append(j._rebuild(limits=[0, end-start, incr]))
-            index = Symbol('%ss%d' % (j.index, i))
-            stmts.append((LocalExpression(Eq(index, j.dim + start), np.int32),
-                          LocalExpression(Eq(index, j.dim - start), np.int32)))
-            mapper[j.dim] = index
+        for t1, t2 in zip(tree, root):
+            start, end, incr = t1.args['limits']
+            index = Symbol('%ss%d' % (t1.index, i))
 
-        # Substitute iteration variables within the folded trees
-        exprs = FindNodes(Expression).visit(otree[-1])
-        replaced = xreplace_indices([j.expr for j in exprs], mapper, only_rhs=True)
-        subs = [j._rebuild(expr=k) for j, k in zip(exprs, replaced)]
+            t1_uindex = (UnboundedIndex(index, start),)
+            t2_uindex = (UnboundedIndex(index, -start),)
+
+            otree.append(t1._rebuild(limits=[0, end-start, incr],
+                                     uindices=t1.uindices + t1_uindex))
+            oroot.append(t2._rebuild(uindices=t2.uindices + t2_uindex))
 
-        handle = Transformer(dict(zip(exprs, subs))).visit(otree[-1])
-        handle = handle._rebuild(nodes=(zip(*stmts)[0] + handle.nodes))
-        processed.append(tuple(otree[:-1]) + (handle,))
+            mapper[t1.dim] = index
 
-        # Temporary arrays can now be moved to the stack
+        # Temporary arrays can now be moved onto the stack
+        exprs = FindNodes(Expression).visit(otree[-1])
         if all(not j.is_Remainder for j in otree):
             shape = tuple(j.bounds_symbolic[1] for j in otree)
-            for j in subs:
-                shape += j.output_function.shape[len(otree):]
-                j.output_function.update(shape=shape, onstack=True)
+            for j in exprs:
+                j_shape = shape + j.output_function.shape[len(otree):]
+                j.output_function.update(shape=j_shape, onstack=True)
+
+        # Substitute iteration variables within the folded trees
+        otree = compose_nodes(otree)
+        replaced = xreplace_indices([j.expr for j in exprs], mapper, only_rhs=True)
+        subs = [j._rebuild(expr=k) for j, k in zip(exprs, replaced)]
```

Comment:
It's always been that way, that's the only main reason. The reason was that ``_rebuild`` isn't an action or a property of an Iteration/Expression tree node, but rather a feature that all nodes of the hierarchy have, so I was simply trying to distinguish that. Might change in a future cleanup PR